src/ucs/memory/memtype_cache.c

```diff
@@ -113,20 +113,48 @@ UCS_PROFILE_FUNC_VOID(ucs_memtype_cache_update_internal,
     start = ucs_align_down_pow2((uintptr_t)address,        UCS_PGT_ADDR_ALIGN);
     end   = ucs_align_up_pow2  ((uintptr_t)address + size, UCS_PGT_ADDR_ALIGN);
 
+    if (action == UCS_MEMTYPE_CACHE_ACTION_SET_MEMTYPE) {
+        /* try to find regions that are contiguous and instersected
+         * with current one */
+        search_start = start - 1;
+        search_end   = end + 1;
+    } else {
+        /* try to find regions that are instersected with current one */
+        search_start = start;
+        search_end   = end;
+    }
+
     pthread_rwlock_wrlock(&memtype_cache->lock);
 
     /* find and remove all regions which intersect with new one */
-    ucs_pgtable_search_range(&memtype_cache->pgtable, start, end - 1,
+    ucs_pgtable_search_range(&memtype_cache->pgtable, search_start, search_end,
                              ucs_memtype_cache_region_collect_callback,
                              &region_list);
-    ucs_list_for_each(region, &region_list, list) {
+    ucs_list_for_each_safe(region, tmp, &region_list, list) {
+        if (action == UCS_MEMTYPE_CACHE_ACTION_SET_MEMTYPE) {
+            if (region->mem_type == mem_type) {
+                /* merge current region with overlapping or adjacent regions
+                 * of same memory type */
+                start = ucs_min(start, region->super.start);
+                end   = ucs_max(end, region->super.end);
+            } else if ((region->super.end < start) ||
+                       (region->super.start >= end)) {
```

Comment:
is this situation possible?

Reply:
yes, I checked it
let's suppose that
1. `[0x100 .. 0x200]` - LAST
`[0x200 .. 0x300]` - CUDA (new region)
or
`[0x100 .. 0x200]` - CUDA (new region)
`[0x200 .. 0x300]` - LAST
they won't be merged since they have different memtypes, and we should save `[0x100 .. 0x200]` (or `[0x200 .. 0x300]`) in the Cache 
2. `[0x100 .. 0x200]` - LAST
`[0x150 .. 0x300]` - CUDA (new region)
it will be updated as:
`[0x100 .. 0x150]` - LAST
`[0x150 .. 0x300]` - CUDA

Reply:
ok